2022/Q2/artifacts/pos-model/PoS-model-cubic.md

```diff
@@ -0,0 +1,517 @@
+    make sense for delbonds.
+    conclusion: intended
+  - https://github.com/informalsystems/partnership-heliax/issues/7
+    there is a problem with unbonding, slashing and total_deltas becoming negative
+    conclusion: it is an actual issue, unresolved
+*/
+//This function is called by transactions tx_unbond, tx_undelegate and tx_redelegate
+func unbond(validator_address, delegator_address, amount)
+{
+  //disallow unbonding if the validator is frozen
+  if (validators[validator_address].frozen == false) then
+    //compute total bonds from delegator to validator
+    var delbonds = compute_total_from_deltas(bonds[delegator_address][validator_address].deltas, cur_epoch + unbonding_length)
+    //check if there are enough bonds
+    //this serves to check that there are bonds (in the docs) and that these are greater than the amount we are trying to unbond
+    if (delbonds >= amount) then
+      //Decrement bond deltas and create unbonds
+      var remain = amount
+      var epoch_counter = cur_epoch + unbonding_length + 1
+      while remain > 0 do
+        epoch_counter = max{epoch | bonds[delegator_address][validator_address].deltas[epoch] > 0 && epoch < epoch_counter}
+        var bond = bonds[delegator_address][validator_address].deltas[epoch_counter]
+        if bond > remain then var unbond_amount = remain
+        else var unbond_amount = bond
+        unbonds[delegator_address][validator_address].deltas[(epoch_counter,cur_epoch+unbonding_length)] += unbond_amount
+        remain -= unbond_amount
+      bonds[delegator_address][validator_address].deltas[cur_epoch+unbonding_length] -= amount
+      //compute new total_deltas and write it at n+unbonding_length
+      var total = read_epoched_field(validators[validator_address].total_deltas, cur_epoch+unbonding_length, 0)
+      validators[validator_address].total_deltas[cur_epoch+unbonding_length] = total - amount
+      //update validator's voting_power, total_voting_power and validator_sets at n+unbonding_length
+      power_before = read_epoched_field(validators[validator_address].voting_power, cur_epoch+unbonding_length, 0)
+      power_after = update_voting_power(validator_address, cur_epoch+unbonding_length)
+      update_total_voting_power(cur_epoch+unbonding_length)
+      update_validator_sets(validator_address, cur_epoch+unbonding_length, power_before, power_after)
+}
+```
+
+```go
+//This function is called by transactions tx_withdraw_unbonds_validator and tx_withdraw_unbonds_delegator
+func withdraw(validator_address, delegator_address)
+{
+  var delunbonds = {<start,end,amount> | amount = unbonds[delegator_address][validator_address].deltas[(start, end)] > 0 && end <= cur_epoch }
+  //substract any pending slash before withdrawing
+  forall (<start,end,amount> in selfunbonds) do
+    var amount_after_slashing = amount
+    forall (slash in slashes[validator_address] s.t. start <= slash.epoch && slash.epoch <= end)
```

Comment:
So, what would happen if validator `validator_address`:

* doesn't have `slashes` (meaning, no punishment against it was processed -- yet -- as part of `end_of_epoch`)
* there are items in `enqueued_slashes` (for an epoch `e > cur_epoch`) that refer to `validator_address`

How is this case handled?

Reply:
I see. If there is no TODO in the model's text, please add one, as this has to be dealt with to have a correct model

Reply:
I have put the link to this comment above the `withdraw` function